api/test.go

```diff
@@ -103,7 +103,9 @@ func WaitForEvent(t *testing.T, client *Client, taskName string, start time.Time
 					continue
 				}
 				mostRecent := task.Events[0]
-				if mostRecent.EndTime.After(start) {
+				if mostRecent.StartTime.After(start) && mostRecent.EndTime.After(start) {
+					// start is the time before a trigger occurs, so this checks if the
+					// most recent event started and completed after the trigger
```

Comment:
Thanks for looking into this! Just want to make sure I understand the cases that this fixes. Is the 'waiting on the wrong event' typically the WaitForEvent after a cli? I can imagine that the cli returns only after the event is stored so the WaitForEvent isn't waiting for the right event. Or is it maybe a different scenario?

I was also curious what your thoughts on if it would be possible to use it in a pattern like:
```
api.WaitForEvent(t, cts, taskName, time.Now(), defaultWaitForEvent) // call Wait ahead
service := testutil.TestService{ID: "api-1", Name: "api"}
testutils.RegisterConsulService(t, srv, service, testutil.HealthPassing,
 	defaultWaitForRegistration)
```
Asking mostly because I've found the e2e tests visually busy and have been thinking about how to reduce lines and make the tests easier on the eyes to read

Reply:
From the uses, it seems that `WaitForEvent` was a helper to wait for the event status to update when a Consul catalog change occurs that triggers a task and finish executing.

Yeah, the scenario for using `WaitForEvent` after a CLI update.

What I found happening was that `WaitForEvent` would return early because it's possible that the main go-routine would run a task at the same time when the `UpdateTask` would run the same task. Both could write an event for the task. When testing for `UpdateTask`'s event, it could catch the event created by the main go-routine because it may have started prior to trigger and complete after the `time.Now()` when `WaitForEvent` was called. Most noticeable for the test `e2e.TestTasksUpdate`.

I'd be all for reducing lines of code! Though I don't know how the specific code example would work out since `WaitForEvent` is a blocking call. I can play around to see what's up

Reply:
Thanks for the details and example of how the WaitForEvent wasn't working! Really helpful! That makes a lot sense.

🤦 completely forgot that WaitForEvent is blocking. No need to spend time playing around with it unless it's something you're very interested in. I'm good with the changes related to fixing WaitForEvent. I know there's an ongoing thread on the changes for the re-enable bug. So I'm planning to hold-off until we decide what changes y'all want to merge.